lib/snippets.coffee

```diff
@@ -322,15 +330,54 @@ module.exports =
         previousTabStopVisited = true
     previousTabStopVisited
 
+  getStore: (editor) ->
+    EditorStore.findOrCreate(editor)
+
   getExpansions: (editor) ->
-    @editorSnippetExpansions?.get(editor) ? []
+    @getStore(editor).getExpansions()
 
   clearExpansions: (editor) ->
-    @editorSnippetExpansions ?= new WeakMap()
-    @editorSnippetExpansions.set(editor, [])
+    store = @getStore(editor)
+    store.clearExpansions()
+    # There are no more active instances of this expansion, so we should undo
+    # the spying we set up on this editor.
+    store.stopObserving()
+    store.stopObservingHistory()
 
   addExpansion: (editor, snippetExpansion) ->
-    @getExpansions(editor).push(snippetExpansion)
+    @getStore(editor).addExpansion(snippetExpansion)
+
+  textChanged: (editor, event) ->
+    store = @getStore(editor)
+    activeExpansions = store.getExpansions()
+
+    return if activeExpansions.length is 0 or activeExpansions[0].isIgnoringBufferChanges
+
+    @ignoringTextChangesForEditor editor, ->
+      editor.transact ->
+        for expansion in activeExpansions
+          expansion.textChanged(event)
+
+    # Create a checkpoint here to consolidate all the changes we just made into
+    # the transaction that prompted them.
+    @makeCheckpoint(editor)
```

Comment:
If I'm reading this right, you're doing this to group any transformations with a previous transaction because the transaction is already closed by the time you receive the event. But I'm confused as to how the initial checkpoint preceding any changes gets populated.

Incidentally, in 1.25 we're adding a `groupLastChanges` method to the buffer that lets you combine the last two transactions on the undo stack, which might make this easier. But you can't really use it until Atom 1.25 hits stable without breaking this package's tests.

Reply:
Working off of recollection here, but I think this is the deal:

The initial snippet expansion is easy because it's invoking a command. Everything that we do in response to that command is automatically atomic because of the `TextEditor#transact` call.

The checkpoint logic is needed only for typing that happens while a snippet is active but not in response to a command invocation. Consider: If I have a snippet like `$1\n${1/./\\u$1}` (i.e., transforming what I type to all-caps on the following line):

1. I type its abbreviation and expand the snippet.
2. I type `a`.
3. Atom inserts `A` on the following line.

We're observing text editor changes, so we act _in between_ steps 2 and 3, but we have to bundle the changes of steps 2 and 3 into one atom — i.e., we have to create a transaction that includes something that already happened. The way I arrived at to do this was to make the checkpoint after the initial snippet expansion and then make a new one each time we did step 3. And right before I make a new checkpoint I bundle everything together that happened since the last checkpoint.

This took a lot of time to reason through. I wrote a lot of tests around this precisely because I suspected I didn't understand it correctly.

It does sound like `groupLastChanges` would make this easier — it would theoretically allow me to toss out checkpoints altogether and duct-tape the changes i'm making with the user action that prompted them.

Reply:
After re-reading your question, I'm pretty sure all I did was restate the part that you already understood. The shorter answer to your question is “I don't know.” I’ll investigate when I get a chance.

Reply:
> The way I arrived at to do this was to make the checkpoint after the initial snippet expansion and then make a new one each time we did step 3

Actually I think this answered my question perfectly.

Reply:
Well, after my long comment I skimmed the source again and can't 100% guarantee that the code does what I claimed it did in the part you quoted. 🤷‍♂️ I can't find the part where I make the initial checkpoint; the `makeCheckpoint` method seems like it only gets called after we set up the listener, not on initial snippet expansion.